### ISSUE TITLE
Clang fails to recognize '\0' as a synonym for 0 when warning

### ISSUE THREAD

**Issue by ChuanqiXu9** [llvm/llvm-project#55402]
Here is the reproducer https://godbolt.org/z/oYxbdaqaY:
```cpp
#include <string.h>

void good(int *a, int n)
{
    memset(a, n, 0);  // warns
}

void bad(int *a, int n)
{
    memset(a, n, '\0');  // fails to warn
}
```

**Comment by efriedma-quic** [comment]
The code intentionally doesn't try to handle arbitrary expressions that evaluate to zero.  I think that's to avoid warning in cases where the correct formula for a length happens to evaluate to zero.

That said, probably makes sense to also handle character literals, particularly in this context.

**Comment by ChuanqiXu9** [comment]
> 

Do you think it makes sense to warn if we could prove an expression would evaluate to zero at constexpr time?

**Comment by efriedma-quic** [comment]
For warnings like this, it's generally a matter of weighing the chance of catching a real mistake, vs the frequency of false positive.

My concern is that someone, for example, allocates a list of N values, them memsets them... but N is a template parameter that's equal to zero for some particular instantiation.  If we throw the full power of constant evaluation at it, the user gets a a likely useless warning, and there isn't any straightforward way to silence it.

If there's a literal zero as the length, we can be pretty confident the user actually screwed up.  Similar if the user writes '\0'.

Given that, the most straightforward thing is probably to just add an explicit check for a CharacterLiteral.

**Comment by ChuanqiXu9** [comment]
Got it. Thanks. I sent a patch to warn for `CharacterLiteral`: https://reviews.llvm.org/D125521

**Comment by ChuanqiXu9** [comment]
Fixed in https://github.com/llvm/llvm-project/commit/3bef90dff64fc717c5d5e33a4d5fb47a4566d04a
